tiledb/cloud/notebook.py

```diff
@@ -35,3 +54,288 @@ def rename_notebook(
+            "Please select a proper namespace (username or organization name).",
+        )
+
+    tries = 1 + retries # 1st + rest
+    while tries > 0:
+        succeeded, tiledb_uri, array_name = _create_notebook_array_retry_helper(
+            storage_path,
+            array_name,
+            namespace,
+            ctx,
+        )
+        if succeeded:
+            return (tiledb_uri, array_name)
+        tries -= 1
+    raise Exception(f"Out of retries uploading to namespace {namespace} array name {array_name}")
+
+
+def _create_notebook_array_retry_helper(
+    storage_path: str,
+    array_name: str,
+    namespace: str,
+    ctx: tiledb.Ctx,
+) -> Tuple[bool, str, str]:
+    """
+    See _create_notebook_array -- exists only for retry logic.
+    :return: tuple of succeeded, tiledb_uri, and array_name
+    """
+
+    try:
+        # The array will be be 1-dimensional with domain of 0 to max uint64. We
+        # use a tile extent of 1024 bytes.
+        dom = tiledb.Domain(
+            tiledb.Dim(
+                name="position",
+                domain=(0, numpy.iinfo(numpy.uint64).max - 1025),
+                tile=1024,
+                dtype=numpy.uint64,
+                ctx=ctx,
+            ),
+            ctx=ctx,
+        )
+
+        schema = tiledb.ArraySchema(
+            domain=dom,
+            sparse=False,
+            attrs=[
+                tiledb.Attr(
+                    name="contents",
+                    dtype=numpy.uint8,
+                    filters=tiledb.FilterList([tiledb.ZstdFilter()]),
+                )
+            ],
+            ctx=ctx,
+        )
+
+        # Goal: tiledb://my_username/s3://my_bucket/my_array
+        # https://docs.tiledb.com/cloud/how-to/arrays/create-arrays
+        tiledb_uri_s3 = "tiledb://" + posixpath.join(namespace, storage_path, array_name)
+
+        # Create the (empty) array on disk.
+        tiledb.DenseArray.create(tiledb_uri_s3, schema)
+        tiledb_uri = "tiledb://" + posixpath.join(namespace, array_name)
+        time.sleep(0.25)
+
+        file_properties = {}
+
+        tiledb.cloud.array.update_info(uri=tiledb_uri, array_name=array_name)
+
+        tiledb.cloud.array.update_file_properties(
+            uri=tiledb_uri,
+            file_type=tiledb.cloud.rest_api.models.FileType.NOTEBOOK,
+            # If file_properties is empty, don't send anything at all.
+            file_properties=file_properties or None,
+        )
+
+        return True, tiledb_uri, array_name
+    except tiledb.TileDBError as e:
+        raise e
+        if "Error while listing with prefix" in str(e):
+            # It is possible to land here if user sets wrong default S3
+            # credentials with respect to default S3 path.
+            raise Exception.web.HTTPError(
+                400, f"Error creating file: {e}. Are your credentials valid?"
+            ) from e
+
+    except Exception as e:
+        raise e
```

Comment:
This probably shouldn't be here anyway (no point in having a catch clause that immediately re-raises and does nothing else) but in the case where you *do* want to re-raise an exception (e.g. after filtering something out), you should use the bare `raise`, which will bubble the exception up without rewriting its call stack etc.